### PR TITLE
Fix kernel naming for Range-based APIs

### PR DESCRIPTION
**Pull request by dmitriy-sobolev**
This fixes an issue with uniqueness of kernel names in Range-based APIs.  

Some Range-bases APIs used to rely on a custom way to generate unique kernel names (look at `__new_kernel_name` function).
The generated kernel names used to be treated as externally provided in the lower layer (`_HasDefaultName` recognizes customizations of kernel names made in the library when using `make_wrapped_policy` in contrast to `__new_kernel_name`). That led to issues with uniqueness of kernel names, because we assume externally provided names are unique even if we do not include types of arguments into a kernel name. You can find a code snippet leading to the error in #507. 

Signed-off-by: Sobolev, Dmitriy <dmitriy.sobolev@intel.com>

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h [hunk 1]
```diff
@@ -39,23 +39,53 @@ namespace __ranges
...
+
+template <typename _Name>
+class __swap1_wrapper
+{
+};
+
+template <typename _Name>
+class __swap2_wrapper
+{
+};
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:67]: This looks rather ugly even for two "wrappers", not even speaking about six.

As an idea (not critical, can be made later), I'd consider these classes to be non-templates. Instead, there might be a universal wrapper template (with two parameters) placed nearby the `make_wrapped_policy` function (which should better be called `make_renamed_policy` BTW) and recognized by `_HasDefaultName`. Seems doable at glance, though we'd need to make sure it works well for the "wrapper over a wrapper over a wrapper" cases.

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h [hunk 2]
```diff
@@ -441,6 +471,16 @@ __pattern_unique(_ExecutionPolicy&& __exec, _Range&& __rng, _BinaryPredicate __p
...
 
+template <typename _Name>
+class __copy_first_wrapper
+{
+};
+
+template <typename _Name>
+class __copy_second_wrapper
+{
+};
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:482]: Does it make sense to make naming more consistent? Here you use "first" and "second" for the names to differ, in other case you use "1" and "2"

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h [hunk 3]
```diff
@@ -649,8 +730,10 @@ __pattern_reduce_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2
...
     //reduce by segment
     oneapi::dpl::__par_backend_hetero::__parallel_for(
-        ::std::forward<_ExecutionPolicy>(__exec), unseq_backend::__brick_reduce_idx<_BinaryOperator>{__binary_op},
-        __result_end, experimental::ranges::views::all_read(__idx), ::std::forward<_Range2>(__values),
+        oneapi::dpl::__par_backend_hetero::make_wrapped_policy<__reduce1_wrapper>(
+            ::std::forward<_ExecutionPolicy>(__exec)),
+        unseq_backend::__brick_reduce_idx<_BinaryOperator>{__binary_op}, __result_end,
+        experimental::ranges::views::all_read(__idx), ::std::forward<_Range2>(__values),
         experimental::ranges::views::all_write(__tmp_out_values))
         .wait();
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:738]: Sometimes we use `oneapi::dpl` "prefix" for referring to a library name (e.g. in the lines 732,733), sometimes we do not (e.g. in the lines 735-737). This is somewhat strange. I thought all our classes (including ranges and bricks) are placed within `namespace oneapi::dpl`; why we skip it sometimes but not all times? Honestly, I think it is excessive when referring to any library name.

I'm not suggesting to immediately change it everywhere, or even in this patch. But coming to an agreement and making the use of library names consistent (in one big effort or with over-time improvements) seems reasonable to me.

**Review by rarutyun — APPROVED** [review]

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h (hunk 1 above)
Comment (dmitriy-sobolev) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:68]: I've created an issue not to forget about the suggestion: #530

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h (hunk 3 above)
Comment (dmitriy-sobolev) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:739]: I've created an issue not to forget about the suggestion: #532

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h [hunk 4]
```diff
@@ -39,23 +39,54 @@ namespace __ranges
...
+{
+};
+
+template <typename _Name>
+class __swap2_wrapper
+{
+};
+
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Function>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy, bool>
```
Comment (SergeyKopienko) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h:71]: this `bool` type looks incorrect.

**Pull request merged**